[Merged by Bors] - feat(topology/uniform_space/uniform_convergence): tendsto_uniformly_on_filter

khwilson opened the pull request:
Currently, mathlib supports several notions of uniform convergence (e.g., tendsto_uniformly_on). These are "global" versions of a more local notion, which we call tendsto_uniformly_on_filter. Specifically, as revealed by tendsto_prod_top_iff is that uniform convergence means convergence on a product filter. So why can't you have a more general filter than a principal filter?

There's no reason you can't! Indeed, if you replace 𝓟 s with 𝓝 x you get a notion of "local uniform convergence" which is enough to prove, e.g., the derivative operator at a point commutes with the pointwise limit.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

khwilson committed:
main commit

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -41,6 +41,13 @@ convergence what a Cauchy sequence is to the usual notion of convergence.
+This definition in and of itself can sometimes be useful, e.g., when studying the local behavior
+of the `Fₙ` near a point, which would typically look like `tendsto_uniformly_on_filter F f p (𝓝 x)`.
+Still, while this may be the "correct" definition (see
+`tendsto_uniformly_on_iff_tendtso_uniformly_on_filter`), it is somewhat unwieldy to work with in
```
```suggestion
`tendsto_uniformly_on_iff_tendsto_uniformly_on_filter`), it is somewhat unwieldy to work with in
```

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -55,21 +62,118 @@ open_locale topological_space classical uniformity filter
+    exact ⟨pb, hpb, pa, hpa, λ x hx y hy, hpapb hy hx⟩, },
+end
+
+lemma eventually_prod_principal_iff {f : ι × α → Prop} {s : set α} :
```
Is this direct from `filter.mem_prod_principal`?

khwilson replied:
Note quite but almost

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -55,21 +62,118 @@ open_locale topological_space classical uniformity filter
+
+variables {α β γ ι : Type*} {p : filter ι} {p' : filter α}
+
+lemma filter.eventually_swap_iff {f : (ι × α) → Prop} : (∀ᶠ (x : ι × α) in (p.prod p'), f x) ↔
```
This should be easy from `filter.eventually_map` and `filter.prod_comm`?

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -55,21 +62,118 @@ open_locale topological_space classical uniformity filter
+    exact ⟨_, ht, _, by simp, htmem⟩, },
+end
+
+lemma filter.prod_le_of_le_of_le {p₁ p₂ : filter ι} {q₁ q₂ : filter α} (hp : p₁ ≤ p₂)
```
[filter.prod_mono](https://leanprover-community.github.io/mathlib_docs/order/filter/basic.html#filter.prod_mono)

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -55,21 +62,118 @@ open_locale topological_space classical uniformity filter
 
 open set filter
 
+section generic
```
all results in this section should be moved to `filter.basic`.

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -55,21 +62,118 @@ open_locale topological_space classical uniformity filter
+
+lemma filter.prod_le_of_right_le (p : filter ι) {p'' : filter α} (hp : p' ≤ p'') :
+  p ×ᶠ p' ≤ p ×ᶠ p'' :=
+filter.prod_le_of_le_of_le rfl.le hp
```
These should be called `mono_left` and `mono_right`

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -55,21 +62,118 @@ open_locale topological_space classical uniformity filter
+
+lemma tendsto_uniformly_on.tendsto_uniformly_on_filter
+  (h : tendsto_uniformly_on F f p s) : tendsto_uniformly_on_filter F f p (𝓟 s) :=
+by rwa ← tendsto_uniformly_on_iff_tendsto_uniformly_on_filter
```
```suggestion
alias tendsto_uniformly_on_iff_tendsto_uniformly_on_filter ↔ 
  tendsto_uniformly_on.tendsto_uniformly_on_filter tendsto_uniformly_on_filter.tendsto_uniformly_on
```
untested, but the other direction might also be useful for dot notation.

fpvandoorn commented on src/topology/uniform_space/uniform_convergence.lean:
```diff
@@ -460,6 +767,7 @@ begin
   obtain ⟨t, ht⟩ := compact_univ.elim_nhds_subcover' (λ k hk, U k) (λ k hk, (hU k).1),
   replace hU := λ (x : t), (hU x).2,
   rw ← eventually_all at hU,
+  -- rw [← principal_univ, eventually_prod_principal_iff],
```
```suggestion
```

fpvandoorn reviewed (commented):
This looks like a good generalization of uniform limits and uniform Cauchy sequences.

In the first section some lemmas need to be moved/removed/cleaned up.

khwilson committed:
replying to floris

Replying to Floris's comments. All done except moving everything to
filter as that will trigger the world's longest build time.

khwilson committed:
migrate lemmas to filter.basic

khwilson committed:
Merge branch 'master' into tendsto-uniformly-on-filter-again

khwilson committed:
migrate to prod

khwilson committed:
forgot one
